File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res=[]
        self.dfs(nums,target,[],res)
        return res
    
    def dfs(self,nums,target,path,res):
        if target<0:
            return #backtracking isme ham add kar rhe hain path main or jab sum bada ho jaa rha ta backtrack kar rhe hai
        if target==0:
            res.append(path)
            return
        for i in range(len(nums)):
            # if target<nums[i]: ham isme value add hi nahi kar rhe, path me jo target se badi ho
            #     continue 
            self.dfs(nums[i:],target-nums[i],path+[nums[i]],res)

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> res= new ArrayList<>();
        dfs(nums,new ArrayList<>(),target, res);
        return res;
    }
    
    public void dfs(int[] nums,List<Integer> path,int target,List<List<Integer>> res){
        if (target == 0){
            res.add(path);
            return;
        }
        for (int i=0;i<nums.length;i++){
            if (nums[i]>target)
                continue;
            List<Integer> c = new ArrayList<>(path);
            c.add(nums[i]);
            dfs(Arrays.copyOfRange(nums,i,nums.length), c, target-nums[i],res);
        }
    }
}